[Foundation]: Add detailed documentation for NSLocalizedString

libsteve opened the pull request:
The existing documentation for the Swift implementation of `NSLocalizedString` doesn't describe its parameters, return value, semantics, and how it relates to localization as a whole.

This PR coalesces information from the Objective-C documentation, Xcode's Help document about exporting localizations, and WWDC presentations on internalization into a descriptive doc-comment.

**Objective-C Documentation**

- [NSLocallizedString][objc-string]
- [NSLocalizedStringFromTable][objc-table]
- [NSLocalizedStringFromTableInBundle][objc-bundle]
- [NSLocalizedStringWithDefaultValue][objc-default]

[objc-string]: https://developer.apple.com/documentation/foundation/nslocalizedstring
[objc-table]: https://developer.apple.com/documentation/foundation/nslocalizedstringfromtable
[objc-bundle]: https://developer.apple.com/documentation/foundation/nslocalizedstringfromtableinbundle
[objc-default]: https://developer.apple.com/documentation/foundation/nslocalizedstringwithdefaultvalue

**WWDC Sessions**

- [WWDC 2017: Localizing with Xcode 9][wwdc-2017-401]
- [WWDC 2018: New Localization Workflows in Xcode 10][wwdc-2018-404]
- [WWDC 2019: Creating Great Localized Experiences with Xcode 11][wwdc-2019-403]

[wwdc-2017-401]: https://developer.apple.com/videos/play/wwdc2017/401/
[wwdc-2018-404]: https://developer.apple.com/videos/play/wwdc2018/404/
[wwdc-2019-403]: https://developer.apple.com/videos/play/wwdc2019/403/

**Xcode Help**

- [Xcode Help: Export localizations for translation][xcode-help]

[xcode-help]: https://help.apple.com/xcode/mac/11.0/index.html?localePath=en.lproj#/deve7c832cd2

libsteve committed:
[Foundation]: Adding detailed documentation for NSLocalizedString

harlanhaskins commented:
Adding @natecook1000 as a reviewer. Thanks @stevebrun!

harlanhaskins commented on stdlib/public/Darwin/Foundation/Foundation.swift:
```diff
@@ -29,7 +29,140 @@ public let NSNotFound: Int = .max
+/// It should be noted that both localization methods can be used at the same
+/// time, but data from manually managed strings file tables will be overwritten
+/// by Xcode if that table is also used with a call to
+/// `NSLocaliedString(_:tableName:bundle:value:comment:)`.
```
Small typo:

```suggestion
/// `NSLocalizedString(_:tableName:bundle:value:comment:)`.
```

libsteve committed:
Fixed some grammar and changed some wording

natecook1000 commented on stdlib/public/Darwin/Foundation/Foundation.swift:
```diff
@@ -29,7 +29,152 @@ public let NSNotFound: Int = .max
+/// ------------
+///
+/// If having Xcode generate strings files from code isn't desired behavior,
+/// you should call `Bundle.localizedString(forKey:value:table:)` instead.
```
```suggestion
/// call `Bundle.localizedString(forKey:value:table:)` instead.
```

natecook1000 commented on stdlib/public/Darwin/Foundation/Foundation.swift:
```diff
@@ -29,7 +29,152 @@ public let NSNotFound: Int = .max
+///     /* A friendly greeting to the user when the program starts. */
+///     "program-greeting" = "Hello, World!";
+///
+/// **Note:** Although `NSLocalizedString(_:tableName:bundle:value:comment:)`
```
The `Note:` syntax gets picked up when it's the first element of a bullet.
```suggestion
/// - Note: Although `NSLocalizedString(_:tableName:bundle:value:comment:)`
```

natecook1000 commented on stdlib/public/Darwin/Foundation/Foundation.swift:
```diff
@@ -29,7 +29,152 @@ public let NSNotFound: Int = .max
+///     - If `value` is the empty string, `key` is returned.
+///     - If `value` is not the empty string, `value` is returned.
+///
+/// Exporting Localizations
```
For this level of header, we use statements that describe the purpose of the following section. Some suggestions for this and the next 2 sections in this comment:

- Use Xcode to Export Localizations
- Choose Meaningful Keys
- Use Only String Literals

stevebrun replied:
What are your thoughts on using "Restrict Autosynthesized Localizations" for the last section? I've been going back and forth between that and "Manually Managing Strings Files".

natecook1000 commented on stdlib/public/Darwin/Foundation/Foundation.swift:
```diff
@@ -29,7 +29,152 @@ public let NSNotFound: Int = .max
+/// Limitations
+/// -----------
+///
+/// Variables and interpolated strings cannot be passed into `key`, `tableName`,
```
Can you phrase this in terms of you should do? i.e. "Only pass string literals as the `key`, ..." After that, you can mention how Xcode doesn't evaluate string interpolations, so translators/int'l users won't see the right thing, and how to use format strings.

natecook1000 commented on stdlib/public/Darwin/Foundation/Foundation.swift:
```diff
@@ -29,7 +29,152 @@ public let NSNotFound: Int = .max
+///                            + "this other string literal works just fine.",
+///                       comment: "The description of a sample of code.")
+///
+/// Luckily, since comments aren't localized, multi-line string literals can be
```
This is a little too informal…
```suggestion
/// Since comments aren't localized, multi-line string literals can be
```

natecook1000 reviewed (changes requested):
Thanks so much for this contribution, @stevebrun! I have a variety of edits below; overall this looks terrific. 👏👏👏

libsteve committed:
Use headers that describe the section's purpose

libsteve committed:
Use `- note:` syntax

libsteve committed:
Use more formal wording

libsteve committed:
Remove extra information from parameter description

This information belongs in the description. Having it both here and there is redundant.

libsteve committed:
Remove the default value from the parameter description

Default values are now automatically included documentation.

libsteve committed:
`key` must not be the empty string

libsteve committed:
Add reason why `key` can't be empty

libsteve committed:
Describe the behavior of `value` without mentioning its default value

libsteve committed:
Reword returns section without bullets

libsteve committed:
Focus on what to do and why to do it

stevebrun commented on stdlib/public/Darwin/Foundation/Foundation.swift:
```diff
@@ -29,7 +29,152 @@ public let NSNotFound: Int = .max
+///   identified by `key`. This is the prefix of the strings file—a file with
+///   the `.strings` extension—containing the localized values. If `tableName`
+///   is `nil` or the empty string, the `Localizable` table is used. The default
+///   value is `nil`.
```
Awesome!

stevebrun commented:
@natecook1000 Thanks for the feedback! 😄

natecook1000 commented on stdlib/public/Darwin/Foundation/Foundation.swift:
```diff
@@ -29,7 +29,153 @@ public let NSNotFound: Int = .max
+///
+/// - parameter key: An identifying value used to reference a localized string.
+///   `key` must not be the empty string. Values keyed by the empty string will
+///   not be localized.
```
Okay, I see — we use "must not" to mean preconditions, which isn't exactly the behavior here. I think you could change this to something like "Don't pass an empty string as a key. Values keyed by an empty string are not localized."

natecook1000 commented on stdlib/public/Darwin/Foundation/Foundation.swift:
```diff
@@ -29,7 +29,153 @@ public let NSNotFound: Int = .max
+/// generating strings files from code. Using those language features will
+/// result in exported localizations that resemble the code expression instead
+/// of its value. Translators would then translate that exported value—leaving
+/// international users with a localized string containing code.
```
This is great, nice improvement!

stevebrun replied:
Thanks 😊

natecook1000 commented on stdlib/public/Darwin/Foundation/Foundation.swift:
```diff
@@ -29,7 +29,153 @@ public let NSNotFound: Int = .max
+///
+/// Xcode does not evaluate interpolated strings and string variables when
+/// generating strings files from code. Using those language features will
+/// result in exported localizations that resemble the code expression instead
```
There's a verb tense disagreement here between "localizations" and "its value". Maybe this instead?

```suggestion
/// result in an exported localization that resembles the code expression instead
```

natecook1000 commented on stdlib/public/Darwin/Foundation/Foundation.swift:
```diff
@@ -29,7 +29,153 @@ public let NSNotFound: Int = .max
+/// Multi-line string literals are technically supported, but will result in
+/// unexpected behavior during internationalization. A newline will be inserted
+/// before and after the body of text within the string, and translators will
+/// likely preserve those in their internaliations.
```
```suggestion
/// likely preserve those in their internationalizations.
```

natecook1000 commented on stdlib/public/Darwin/Foundation/Foundation.swift:
```diff
@@ -29,7 +29,153 @@ public let NSNotFound: Int = .max
+///     NSLocalizedString("multi-line-string-literal",
+///                       value: """
+///     This multi-line string literal won't work as expected.
+///     An extra newline added to the beginning and end of the string.
```
```suggestion
///     An extra newline is added to the beginning and end of the string.
```

natecook1000 commented on stdlib/public/Darwin/Foundation/Foundation.swift:
```diff
@@ -29,7 +29,153 @@ public let NSNotFound: Int = .max
+/// Since comments aren't localized, multi-line string literals can be safely
+/// used with `comment`.
+///
+/// Restrict Autosynthesized Localizations
```
Maybe "Work With Manually Managed Strings"?

stevebrun replied:
"Mitigate Mayhem With Manually Managed Strings " 😆

"Work With Manually Managed Strings" sounds much better. I'll go with that one. Thanks!

natecook1000 commented on stdlib/public/Darwin/Foundation/Foundation.swift:
```diff
@@ -29,7 +29,153 @@ public let NSNotFound: Int = .max
+///
+/// - note: Although `NSLocalizedString(_:tableName:bundle:value:comment:)`
+/// and `Bundle.localizedString(forKey:value:table:)` can be used in a project
+/// at the same time, data from a manually managed strings files will be
```
```suggestion
/// at the same time, data from manually managed strings files will be
```

natecook1000 commented on stdlib/public/Darwin/Foundation/Foundation.swift:
```diff
@@ -29,7 +29,153 @@ public let NSNotFound: Int = .max
+///                                    comment: "A math equation.")
+///     let localizedString = String.localizedStringWithFormat(format, (1 + 1))
+///
+/// Multi-line string literals are technically supported, but will result in
```
```suggestion
/// Multiline string literals are technically supported, but will result in
```
Here and below, too…

libsteve committed:
Fix verb tense disagreement

libsteve committed:
Correct spelling mistakes and typos

libsteve committed:
"Must" must not be used outside a precondition

libsteve committed:
Fix wording for manual string management header

libsteve committed:
Capitalize the w in with

stevebrun commented:
@natecook1000 Thanks for catching those typos! Do you have any other suggestions?

natecook1000 committed:
Revert "Capitalize the w in with"

This was my error — with should be lowercase in titles.

natecook1000 commented:
@swift-ci Please smoke test and merge

natecook1000 commented:
Looks great @stevebrun, thanks for the updates!

natecook1000 commented:
@swift-ci Please smoke test and merge
